### PR TITLE
Rename data key to data marker [info]

### PR DESCRIPTION
**Pull request by robertbastian**
#4991

### CODE REVIEW

File: provider/core/src/key.rs [hunk 1]
```diff
@@ -157,267 +148,105 @@ impl AsULE for DataKeyHash {
...
+    tagged: &'static str,
+    hash: DataMarkerPathHash,
 }
 
-impl PartialEq for DataKey {
+impl PartialEq for DataMarkerPath {
     #[inline]
     fn eq(&self, other: &Self) -> bool {
-        self.hash == other.hash && self.path == other.path && self.metadata == other.metadata
+        self.hash == other.hash
```
Comment (sffc) [inline provider/core/src/key.rs:180]: Issue: Please add the self.path back to the cmp function

File: provider/core/src/key.rs [hunk 2]
```diff
@@ -463,213 +292,286 @@ impl DataKey {
...
+
+/// Used for loading data from a dynamic ICU4X data provider.
+///
+/// A data marker is tightly coupled with the code that uses it to load data at runtime.
+/// Executables can be searched for `DataMarkerInfo` instances to produce optimized data files.
+/// Therefore, users should not generally create DataMarkerInfo instances; they should instead use
+/// the ones exported by a component.
+#[derive(Copy, Clone, PartialEq, Eq)]
+#[allow(clippy::exhaustive_structs)] // TODO
+pub struct DataMarkerInfo {
```
Comment (sffc) [inline provider/core/src/key.rs:344]: Issue: Should be non-exhaustive

(To keep it non-exhaustive, keep using the struct)

**Comment by robertbastian** [comment]
Sorry @all, I cannot remove reviewers because GitHub is throwing 404s. I only want a review from @sffc

File: components/locale/src/expander.rs [hunk 3]
```diff
@@ -650,20 +650,24 @@ mod tests {
...
+            if self.markers.contains(&marker) {
+                return Err(DataErrorKind::MissingDataMarker.with_str_context("rejected"));
             }
 
             let l = crate::provider::Baked::SINGLETON_LOCID_TRANSFORM_LIKELYSUBTAGS_L_V1;
             let ext = crate::provider::Baked::SINGLETON_LOCID_TRANSFORM_LIKELYSUBTAGS_EXT_V1;
             let sr = crate::provider::Baked::SINGLETON_LOCID_TRANSFORM_LIKELYSUBTAGS_SR_V1;
 
-            let payload = if key.hashed() == LikelySubtagsV1Marker::KEY.hashed() {
+            let payload = if marker.path.hashed() == LikelySubtagsV1Marker::INFO.path.hashed() {
```
Comment (sffc) [inline components/locale/src/expander.rs:670]: Thought we should drop the old keys (not in this pr)

File: provider/adapters/src/fork/mod.rs [hunk 4]
```diff
@@ -6,12 +6,12 @@
 //!
 //! # Types of Forking Providers
 //!
-//! ## Key-Based
+//! ## marker-Based
 //!
```
Comment (sffc) [inline provider/adapters/src/fork/mod.rs:10]: Capitalization

**Review by sffc — COMMENTED** [review]
Through provider/blob. Good so far

File: components/datetime/tests/data/gen.sh [hunk 5]
```diff
@@ -1,6 +1,6 @@
 #!/bin/sh
 cargo run -p icu_datagen -- \
---keys "datetime/gregory/datelengths@1" "datetime/gregory/datesymbols@1" "datetime/timelengths@1" "datetime/timesymbols@1" "decimal/symbols@1" "time_zone/formats@1" "time_zone/specific_short@1" \
+--markers "datetime/gregory/datelengths@1" "datetime/gregory/datesymbols@1" "datetime/timelengths@1" "datetime/timesymbols@1" "decimal/symbols@1" "time_zone/formats@1" "time_zone/specific_short@1" \
```
[the remark this answers is not in the document]
Comment (robertbastian) [reply components/datetime/tests/data/gen.sh:3]: hmm?

File: components/experimental/tests/transliterate/data/baked/macros.rs [hunk 6]
```diff
@@ -26,33 +26,12 @@ macro_rules! __make_provider {
 #[doc(inline)]
 pub use __make_provider as make_provider;
 #[macro_use]
-#[path = "macros/normalizer_comp_v1.rs.data"]
-mod normalizer_comp_v1;
```
[the remark this answers is not in the document]
Comment (robertbastian) [reply components/experimental/tests/transliterate/data/baked/macros.rs]: why indeed

File: provider/adapters/src/fork/mod.rs [hunk 7]
```diff
@@ -6,7 +6,7 @@
...
 //! # Types of Forking Providers
 //!
-//! ## marker-Based
+//! ## Mcararker-Based
```
Comment (sffc) [inline provider/adapters/src/fork/mod.rs:9]: Mcararker? Caramaraker? Ramcarkerer?

Comment (robertbastian) [reply provider/adapters/src/fork/mod.rs:9]: lol how?! I've noticed that VSCode sometimes corrupts files on large rename tasks, but where does the c from 🤔

File: provider/core/src/lib.rs [hunk 8]
```diff
@@ -220,15 +219,15 @@ pub mod prelude {
...
+    pub use crate::DataMarker;
     #[doc(no_inline)]
-    pub use crate::DataKeyHash;
+    pub use crate::DataMarkerAttributes;
     #[doc(no_inline)]
-    pub use crate::DataLocale;
+    pub use crate::DataMarkerInfo;
     #[doc(no_inline)]
-    pub use crate::DataMarker;
+    pub use crate::DataMarkerPathHash;
```
Comment (sffc) [inline provider/core/src/lib.rs:230]: These entries were re-ordered. For my own record:

| Old | New |
|---|---|
| DataKey | DataMarkerInfo |
| DataKeyAttributes | DataMarkerAttribute |
| DataKeyHash | DataMarkerPathHash |
| DataLocale | DataLocale |
| DataMarker | DataMarker |

File: provider/core/src/lib.rs [hunk 9]
```diff
@@ -154,16 +154,15 @@ pub mod serde;
...
 pub use crate::data_provider::DynamicDataProvider;
 pub use crate::error::DataError;
 pub use crate::error::DataErrorKind;
-pub use crate::key::DataKey;
-pub use crate::key::DataKeyHash;
-pub use crate::key::DataKeyMetadata;
-pub use crate::key::DataKeyPath;
-pub use crate::request::DataKeyAttributes;
+pub use crate::key::DataMarkerInfo;
+pub use crate::key::DataMarkerPath;
```
Comment (sffc) [inline provider/core/src/lib.rs:162]: Should DataMarkerPath go in the prelude?

**Review by sffc — COMMENTED** [review]
Down through provider/core

File: provider/core/src/lib.rs (hunk 9 above)
Comment (robertbastian) [reply provider/core/src/lib.rs:162]: I think that's something we can consider later, if we can change the datagen API to accept paths instead of infos.

Comment (sffc) [reply provider/core/src/lib.rs:162]: Seems like we should have DataMarkerPath, DataMarkerPath+DataMarkerPathHash, or neither? Right now we have DataMarkerPathHash but not DataMarkerPath in the prelude

Comment (robertbastian) [reply provider/core/src/lib.rs:162]: I'll remove the hash for now, I think that's specific enough to not be in the prelude

**Review by sffc — APPROVED** [review]

**Pull request merged**